Change the variable name continuationHistory in Stack to pieceToHistory.

vdbergh opened the pull request:
The name continuationHistory is confusing since the type of the variable is actually PieceToHistory* and there is also a type ContinuationHistory.

We also add a comment that the (now renamed) pieceToHistory instance variable points to an appropriate entry in Thread::continuationHistory (which is of type ContinuationHistory[2][2]).

No functional change.

ddugovic commented:
This isn't your fault, but since this code has no comment I could never understand what it does.

vdbergh commented:
Using sensible variable names should already help :)

It also took me some time to understand that << when applied to a StatEntry has nothing to do with left shift.  If overloading has to be used then I guess I would have preferred += (the operation is some sort of clipped addition).

snicolet commented:
I agree that this data structure of code is not the easiest to understand in the codebase, because the idea is subtle.

A long time ago we had a simple, flat array called history for storing the stats of each square of the board, like all the other chess engines. Then Stefan Geschwentner invented the counter move history (which was a huge idea worth more than 10 Elo by itself), which basically puts the previous move by the opponent into the equation to add depth to the stats. The new array was doubly indexed by the previous moves and the current move (so at plies n-1 and n), saying  something like "if the opponent has just played this move, then the average stats for each of our moves are the following".

Then Stefan again generalized his idea by adding stats indexed by move at ply n-2, n-4, etc. This allowed Stockfish to store stats for (partial) plans rather than individual moves or move pairs, and is, I think, a large part of why SF move ordering is so efficient.

However the initial implementation was a bit confusing, accessing the stack at different levels in the movePicker to access different pointers called `counterMove` at different levels, when you would think that a counter move is always an answer to the last opponent threat.

Marco changed it in PR 1180, deciding that it was better to put the logic of accessing the different levels of stack in search.cpp, and passing the various arrays as parameters to the MovePicker constructor. It was also in PR 1180 that the name ContinuationHistory was chosen, since it sounded like an improvement to generalize both concepts of (1) counter move against the last opponent threat and (2) follow-up moves for our moves at plies n-2, n-4,...

See https://github.com/official-stockfish/Stockfish/pull/1180 for the discussion at that time.

vdbergh commented:
@snicolet I understand the idea and I have no objection to the name of _type_ ContinuationHistory. I have an objection to the name of the _instance variable_ continuationHistory. The lattter is not of type ContinuationHistory but of type PieceToHistory. This was very confusing to me.

Note that there is another instance variable in the code which is called continuationHistory and which is indeed of type ContinuationHistory. Of course I left that one alone.

vdbergh commented:
I added a comment that the (now renamed) pieceToHistory instance variable points to an appropriate entry in Thread::continuationHistory (which is of type ContinuationHistory[2][2]).
